WordPress/Classes/Utility/ContextManager.swift

```diff
@@ -0,0 +1,239 @@
+        assert(storeURL.isFileURL)
+
+        self.modelName = modelName
+        self.storeURL = storeURL
+        self.persistentContainer = Self.createPersistentContainer(storeURL: storeURL, modelName: modelName)
+
+        super.init()
+
+        mainContext.automaticallyMergesChangesFromParent = true
+        mainContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        NullBlogPropertySanitizer(context: mainContext).sanitize()
+    }
+
+    public func newDerivedContext() -> NSManagedObjectContext {
+        let context = persistentContainer.newBackgroundContext()
+        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        return context
+    }
+
+    @objc(performAndSaveUsingBlock:)
+    public func performAndSave(_ block: @escaping (NSManagedObjectContext) -> Void) {
+        let context = newDerivedContext()
+        context.performAndWait {
+            block(context)
+
+            self.save(context, andWait: true, withCompletionBlock: nil)
+        }
+    }
+
+    @objc(performAndSaveUsingBlock:completion:)
+    public func performAndSave(_ block: @escaping (NSManagedObjectContext) -> Void, completion: @escaping () -> Void) {
+        let context = newDerivedContext()
+        context.perform {
+            block(context)
+
+            self.save(context, andWait: false, withCompletionBlock: completion)
+        }
+    }
+
+    @objc
+    public func saveContextAndWait(_ context: NSManagedObjectContext) {
+        save(context, andWait: true, withCompletionBlock: nil)
+    }
+
+    @objc(saveContext:)
+    public func save(_ context: NSManagedObjectContext) {
+        save(context, andWait: false, withCompletionBlock: nil)
+    }
+
+    @objc(saveContext:withCompletionBlock:)
+    public func save(_ context: NSManagedObjectContext, withCompletionBlock completion: @escaping () -> Void) {
+        save(context, andWait: false, withCompletionBlock: completion)
+    }
+}
+
+// MARK: - Private methods
+
+private extension ContextManager {
+    static var localDatabasePath: URL {
+        guard let url = try? FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true) else {
+            fatalError("Failed to find the document directory")
+        }
+
+        return url.appendingPathComponent("WordPress.sqlite")
+    }
+
+    func save(_ context: NSManagedObjectContext, andWait wait: Bool, withCompletionBlock completionBlock: (() -> Void)?) {
+        let block: () -> Void = {
+            self.internalSave(context)
+            DispatchQueue.main.async {
+                completionBlock?()
+            }
+        }
+
+        /// Ensure that the `context`'s concurrency type is not `confinementConcurrencyType`, since it will crash if `perform` or `performAndWait` is called.
+        guard context.concurrencyType == .mainQueueConcurrencyType || context.concurrencyType == .privateQueueConcurrencyType else {
```

Comment:
The comment says we want to check that the context `!= x`. The code checks that `== y || == z`. 

The difference might be important. I think it would be good to keep code and comment consistent.

Also, nitpick, `///` will result in a documentation header comment, but that doesn't apply for this kind of comment since it's not on a definition but on a guard statement.

Reply:
The original PR chose to not reference the deprecated `confinementConcurrencyType` to [avoid a warning](https://github.com/wordpress-mobile/WordPress-iOS/pull/19765#discussion_r1044740441).